File: LeetCode/266-palindrome-permutation/266-palindrome-permutation.py
class Solution:
    # def canPermutePalindrome(self, s: str) -> bool:
    #     oddChars = set()

    #     for c in s:
    #         if c in oddChars:
    #             oddChars.remove(c)
    #         else:
    #             oddChars.add(c)

    #     return len(oddChars) <= 1

    def canPermutePalindrome(self, s: str) -> bool:
        char_dict = {}
        for c in s:
            char_dict[c] = char_dict.get(c, 0)+1

        odd_count = 0
        for v in char_dict.values():
            if v % 2 == 1:
                odd_count += 1
                if odd_count > 1:
                    return False

        return True


if __name__ == '__main__':
    sol = Solution()
    inputs = ["code", "aab", "carerac"]
    outputs = [False, True, True]

    for pos in range(len(outputs)):
        ans = sol.canPermutePalindrome(inputs[pos])
        if ans != outputs[pos]:
            print(f'{inputs[pos]}, {ans}, {outputs[pos]}')

File: LeetCode/266-palindrome-permutation/266-palindrome-permutation.go
package main

import "fmt"

func canPermutePalindrome(s string) bool {
	cnt := [26]int{}
	for _, c := range s {
		cnt[c-'a']++
	}
	odd := 0
	for _, x := range cnt {
		odd += x & 1
	}
	return odd < 2
}

func main() {
	inputs := []string{"code", "aab", "carerac"}
	outputs := []bool{false, true, true}
	var ans bool

	for pos := 0; pos < len(outputs); pos++ {

		ans = canPermutePalindrome(inputs[pos])
		if ans != outputs[pos] {
			fmt.Println(inputs[pos], outputs[pos], ans)
		}
	}
}
